styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -183,6 +190,75 @@
     this.forkJoinPool = new ForkJoinPool(REQUEST_CONCURRENCY);
   }
 
+  /**
+   * Bootstrap active worflow instance index. This cannot run concurrently with the scheduler and must only
+   * be called once before the scheduler starts.
+   */
+  void indexActiveWorkflowInstances() {
+    // XXX: this listing is not strongly consistent
+    final List<Entity> indexEntries = StreamUtil.stream(
+        datastore.run(Query.newKeyQueryBuilder().setKind(KIND_ACTIVE_WORKFLOW_INSTANCE).build()))
+        .map(Key::getName)
+        .map(wfiKey -> activeWorkflowInstanceIndexShardEntryKey(datastore.newKeyFactory(), wfiKey))
+        .map(indexEntryKey -> Entity.newBuilder(indexEntryKey).build())
+        .collect(toList());
+
+    for (List<Entity> batch : Lists.partition(indexEntries, MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH)) {
+      datastore.put(batch.toArray(new Entity[0]));
+    }
+  }
+
+  /**
+   * This indexing can run concurrently with the scheduler operations as it is using transactions to ensure it
+   * consistently updates the index, but that also makes it too slow to do the initial bootstrap.
+   *
+   */
+  void concurrentlyIndexActiveWorkflowInstances() {
+    // Weak listing, but we'll eventually read _everything_, in due time...
+    final List<Entity> expectedIndexEntries = StreamUtil.stream(
+        datastore.run(Query.newKeyQueryBuilder().setKind(KIND_ACTIVE_WORKFLOW_INSTANCE).build()))
+        .map(Key::getName)
+        .map(wfiKey -> activeWorkflowInstanceIndexShardEntryKey(datastore.newKeyFactory(), wfiKey))
+        .map(indexEntryKey -> Entity.newBuilder(indexEntryKey).build())
+        .collect(toList());
+
+    // Fetch index entries in batches
+    for (List<Entity> expectedEntryBatch : Lists.partition(expectedIndexEntries, MAX_NUMBER_OF_ENTITIES_IN_ONE_BATCH)) {
+
+      final List<Entity> actualEntryBatch = datastore.fetch(expectedEntryBatch.stream()
```

Comment:
Is the eager loading intended? It feels easier to understand if we do `datastore.get` and then diff the two lists. Or performance is the concern here because List subtraction is slow? If that is the case, it is fine.

Reply:
Yeah, the eager loading was intended. The diffing is happening below.